src/java.base/share/classes/jdk/internal/module/ModuleReferenceImpl.java

```diff
@@ -62,8 +62,18 @@ public class ModuleReferenceImpl extends ModuleReference {
     // ModuleResolution flags
     private final ModuleResolution moduleResolution;
 
-    // cached hash of this module to avoid needing to compute it many times
-    private byte[] cachedHash;
+    // cached hash of this module to avoid needing to compute it many times.
+    // For safety reasons under concurrent updates, we need to wrap the
+    // hash/algo pair into the final wrapper.
+    private CachedHash cachedHash;
+    private final class CachedHash {
+        public CachedHash(byte[] hash, String algorithm) {
+            this.hash = hash;
+            this.algorithm = algorithm;
+        }
+        final byte[] hash;
+        final String algorithm;
```

Comment:
Have you tried make this a record? Also can you move the declaration of cachedHash to after the declaration of the record.

Reply:
Switched to `record`, moved.